File: web/studio/ASC.Web.Studio/UserControls/Common/PollForm/PollForm.cs
/*
 *
 * (c) Copyright Ascensio System Limited 2010-2020
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/


using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using AjaxPro;
using Resources;

namespace ASC.Web.Studio.UserControls.Common.PollForm
{
    public interface IVoteHandler
    {
        bool VoteCallback(string pollID, List<string> selectedVariantIDs, string additionalParams, out string errorMessage);
    }

    [AjaxNamespace("PollFormControl")]
    [DefaultProperty("Name")]
    [ToolboxData("<{0}:PollForm runat=server></{0}:PollForm>")]
    public class PollForm : WebControl
    {
        [Bindable(true)]
        [Category("Appearance")]
        [DefaultValue("")]
        [Localizable(true)]
        public string Name
        {
            get
            {
                var s = (String)ViewState["Name"];
                return s ?? String.Empty;
            }
            set { ViewState["Name"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public bool Answered
        {
            get
            {
                var o = ViewState["Answered"];
                return o != null && (bool)o;
            }
            set { ViewState["Answered"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public bool Singleton
        {
            get
            {
                var o = ViewState["Singleton"];
                return o != null && (bool)o;
            }
            set { ViewState["Singleton"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public Type VoteHandlerType
        {
            get
            {
                var o = ViewState["VoteHandlerType"];
                return (o == null) ? null : (Type)o;
            }
            set { ViewState["VoteHandlerType"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public List<AnswerViarint> AnswerVariants
        {
            get
            {
                var o = ViewState["AnswerVariants"];
                return (o == null) ? null : (List<AnswerViarint>)o;
            }
            set { ViewState["AnswerVariants"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string AdditionalParams
        {
            get
            {
                var o = ViewState["AdditionalParams"];
                return (o == null) ? null : (string)o;
            }
            set { ViewState["AdditionalParams"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string ButtonCSSClass
        {
            get
            {
                var o = ViewState["ButtonCSSClass"];
                return (o == null) ? null : (string)o;
            }
            set { ViewState["ButtonCSSClass"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string StatisticBarCSSClass
        {
            get
            {
                var o = ViewState["StatisticBarCSSClass"];
                return (o == null) ? null : (string)o;
            }
            set { ViewState["StatisticBarCSSClass"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string LiderStatisticBarCSSClass
        {
            get
            {
                var o = ViewState["LiderStatisticBarCSSClass"];
                return ((o == null) ? null : (string)o);
            }
            set { ViewState["LiderStatisticBarCSSClass"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string VariantTextCSSClass
        {
            get
            {
                var o = ViewState["VariantTextCSSClass"];
                return (o == null) ? null : (string)o;
            }
            set { ViewState["VariantTextCSSClass"] = value; }
        }

        [Bindable(true)]
        [Category("Appearance")]
        public string VoteCountTextCSSClass
        {
            get
            {
                var o = ViewState["VoteCountTextCSSClass"];
                return (o == null) ? null : (string)o;
            }
            set { ViewState["VoteCountTextCSSClass"] = value; }
        }

        public string PollID { get; set; }
        public string BehaviorID { get; set; }
        private string _jsObjName;

        public PollForm()
        {
            AnswerVariants = new List<AnswerViarint>();
            VoteCountTextCSSClass = "pollSmallDescribe";
            VariantTextCSSClass = "VariantTextCSSClass";
            LiderStatisticBarCSSClass = "statisticBar liaderBar";
            StatisticBarCSSClass = "statisticBar";
            ButtonCSSClass = "button gray";
        }

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);

            AjaxPro.Utility.RegisterTypeForAjax(GetType());
            Page.RegisterBodyScripts("~/UserControls/Common/PollForm/js/pollform.js");
        }

        protected override void OnPreRender(EventArgs e)
        {
            _jsObjName = String.IsNullOrEmpty(BehaviorID) ? ("__pollForm" + UniqueID) : BehaviorID;

            var script = new StringBuilder();
            script.Append(" window." + _jsObjName + " = new VotingPollPrototype('" + _jsObjName + "','" + VoteHandlerType.AssemblyQualifiedName + "','" + PollID + "','" + UserControlsCommonResource.EmptySelect + "','" + StatisticBarCSSClass + "','" + LiderStatisticBarCSSClass + "','" + VariantTextCSSClass + "','" + VoteCountTextCSSClass + "','" + AdditionalParams + "'); ");

            foreach (var variant in AnswerVariants)
                script.Append(_jsObjName + ".RegistryVariant(new AnswerVariantPrototype('" + variant.ID + "','" + ReplaceSingleQuote(HttpUtility.HtmlEncode(variant.Name)) + "'," + variant.VoteCount + ")); ");

            Page.RegisterInlineScript(script.ToString());
        }

        protected override void RenderContents(HtmlTextWriter output)
        {
            output.Write(RenderPollForm());
        }

        private string RenderPollForm()
        {
            var sb = new StringBuilder();

            sb.Append("<div>");
            sb.AppendFormat(@"<div id=""__pollForm_{0}_result""></div>", PollID);
            sb.AppendFormat(@"<div id=""__pollForm_{0}_voteBox"">", PollID);

            if (!Answered)
            {
                foreach (var variant in AnswerVariants)
                {
                    sb.Append(@"<div style=""padding:6px 0px"">");

                    sb.AppendFormat(@"<input name=""__pollForm_{0}_av"" value=""{1}"" id=""__pollForm_{0}_av_{1}"" style=""float: left; margin: 0 10px 0 0;"" type=""{2}""/><label for=""__pollForm_{0}_av_{1}"">{3}</label>",
                                    PollID, variant.ID, (Singleton ? "radio" : "checkbox"), ReplaceSingleQuote(HttpUtility.HtmlEncode(variant.Name)));

                    sb.Append("</div>");
                }
            }
            else
            {
                sb.Append(RenderAnsweredVariants(AnswerVariants, StatisticBarCSSClass, LiderStatisticBarCSSClass, VariantTextCSSClass));
            }


            if (!Answered)
                sb.AppendFormat(@"<div id=""__pollForm_{0}_answButtonBox"" class=""small-button-container""><a class=""{3}"" href=""javascript:{1}.Vote();"">{2}</a></div>",
                                PollID, _jsObjName, UserControlsCommonResource.VoteButton, ButtonCSSClass);

            sb.Append(@"</div>");
            sb.Append(@"</div>");
            return sb.ToString();
        }

        private string RenderAnsweredVariants(List<AnswerViarint> variants, string statBarCSSClass, string liderBarCSSClass, string variantNameCSSClass)
        {
            const float width = 70;
            double k = 0;
            double userCount;
            double max = 0, fullCount = 0;

            foreach (var variant in variants)
            {
                userCount = variant.VoteCount;
                fullCount += userCount;
                if (max < userCount)
                    max = userCount;
            }

            if (max != 0)
                k = width/max;

            var sb = new StringBuilder();

            foreach (var variant in variants)
            {
                userCount = variant.VoteCount;

                sb.AppendFormat(@"<div class=""{0}"">", variantNameCSSClass);
                sb.AppendFormat("<b>{0}</b>", HttpUtility.HtmlEncode(variant.Name));
                sb.Append(@"<span class=""splitter""></span>");
                sb.AppendFormat(@"<span class=""gray-text"">{0} ({1}%)</span>", userCount, fullCount != 0 ? Math.Round((userCount*100)/fullCount) : 0);
                sb.Append("</div>");
                sb.AppendFormat(@"<div class=""{1}"" style=""width:{0}%;"">&nbsp;</div>", Math.Round(k*userCount), max == userCount ? liderBarCSSClass : statBarCSSClass);
                sb.Append(@"<div style=""clear:both;"">&nbsp;</div>");
            }
            sb.AppendFormat(@"<div class=""{2}"">{0}: {1}</div>", UserControlsCommonResource.AllVoting, fullCount, variantNameCSSClass);

            return sb.ToString();
        }

        public class AnswerViarint
        {
            public string ID { get; set; }
            public string Name { get; set; }
            public int VoteCount { get; set; }
        }

        private class VoteResult
        {
            public string Message { get; set; }
            public string Success { get; set; }
            public string HTML { get; set; }
            public string PollID { get; set; }
        }

        [AjaxMethod(HttpSessionStateRequirement.ReadWrite)]
        public object Vote(string voteHandlerTypeName, string pollID, string variantIDs, List<AnswerViarint> allVariants,
                           string statBarCSSClass, string liderBarCSSClass, string variantNameCSSClass, string voteCountCSSClass,
                           string additionalParams)
        {
            var result = new VoteResult { PollID = pollID };

            var voteHandler = (IVoteHandler)Activator.CreateInstance(Type.GetType(voteHandlerTypeName));
            var selectedVariantIDs = new List<string>();

            selectedVariantIDs.AddRange(variantIDs.Split(new char[] { ',' }, StringSplitOptions.RemoveEmptyEntries));

            string errorMessage;
            if (voteHandler.VoteCallback(pollID, selectedVariantIDs, additionalParams, out errorMessage))
            {
                result.Success = "1";
                foreach (var var in allVariants)
                {
                    var selectedID = selectedVariantIDs.Find(id => String.Equals(id, var.ID, StringComparison.InvariantCultureIgnoreCase));
                    if (!String.IsNullOrEmpty(selectedID))
                        var.VoteCount++;
                }

                allVariants.ForEach(v => v.Name = HttpUtility.HtmlDecode(v.Name));

                result.HTML = RenderAnsweredVariants(allVariants, statBarCSSClass, liderBarCSSClass, variantNameCSSClass);
            }
            else
            {
                result.Success = "0";
                result.Message = HttpUtility.HtmlEncode(errorMessage);
            }

            return result;
        }

        public static string ReplaceSingleQuote(string str)
        {
            return str == null ? null : str.Replace("'", "′");
        }
    }
}

File: web/studio/ASC.Web.Studio/UserControls/Common/PollForm/js/pollform.js
/*
 *
 * (c) Copyright Ascensio System Limited 2010-2020
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/


AnswerVariantPrototype = function (id, name, voteCount) {
    this.ID = id;
    this.Name = name;
    this.VoteCount = voteCount;
};

VotingPollPrototype = function (varName, voteHandlerType, pollID, emptySelectText,
                                statBarCSSClass, liderBarCSSClass, variantNameCSSClass, voteCountCSSClass,
                                additionalParams) {
    this.VarName = varName;
    this.VoteHandlerType = voteHandlerType;
    this.PollID = pollID;
    this.EmptySelect = emptySelectText;
    this.AnswerVariants = new Array();
    this.AdditionalParams = additionalParams;

    this.StatBarCSSClass = statBarCSSClass;
    this.LiderBarCSSClass = liderBarCSSClass;
    this.VariantNameCSSClass = variantNameCSSClass;
    this.VoteCountCSSClass = voteCountCSSClass;

    this.RegistryVariant = function (variant) {
        for (var i = 0; i < this.AnswerVariants.length; i++) {
            if (this.AnswerVariants[i].ID == variant.ID)
                return;
        }
        this.AnswerVariants.push(variant);
    };

    this.Vote = function () {
        var ids = new String();
        jq.each(jq("input:checked[name='__pollForm_" + this.PollID + "_av']"), function (i, n) {
            if (i != 0)
                ids += ",";

            ids += jq(this).val();
        });

        jq('#__pollForm_' + this.PollID + '_result').html('');

        if (ids == '') {
            jq('#__pollForm_' + this.PollID + '_result').html('<div class="errorBox">' + this.EmptySelect + '</div>');
            return;
        }

        AjaxPro.onLoading = function (b) {
            if (b)
                LoadingBanner.showLoaderBtn('#__pollForm_' + this.PollID + '_voteBox');
            else
                LoadingBanner.hideLoaderBtn('#__pollForm_' + this.PollID + '_voteBox');
        };

        PollFormControl.Vote(this.VoteHandlerType, this.PollID, ids, this.AnswerVariants,
            this.StatBarCSSClass, this.LiderBarCSSClass, this.VariantNameCSSClass, this.VoteCountCSSClass,
            this.AdditionalParams, function (result) {
                var res = result.value;
                if (res.Success == '1') {
                    jq('#__pollForm_' + res.PollID + '_answButtonBox').hide();
                    jq('#__pollForm_' + res.PollID + '_voteBox').html(res.HTML);
                } else {
                    jq('#__pollForm_' + res.PollID + '_result').html('<div class="errorBox">' + res.Message + '</div>');
                }
            });
    };
};
PollMaster = new function () {
    this.AddAnswerVariant = function (uniqueID, variantClass, labelClass, editClass) {
        var items = jq('#__poll_' + uniqueID + '_qbox').children();
        var numb = items.length + 1;
        var leftMargin = 10;
        if (numb > 9)
            leftMargin = 3;
        var answerId = "__poll_" + uniqueID + "_q" + numb;
        var sb = '<div class="' + variantClass + '"><label for="__poll_' + uniqueID + '_q' + numb + '" style="float:left;" class="' + labelClass + '">' +
            jq('#__poll_' + uniqueID + '_variantCaption').val() + ' ' + numb +
            ':</label><span class="poll_remove_span_' + numb + '" onclick="PollMaster.RemoveAnswerVariant(this)"></span><input id="' + answerId + '" name="q' + numb + '" type="text" maxlength="100" class="' + editClass + '" style="margin:2px 0 0 ' + leftMargin + 'px;  float:left; width:500px"/>' +
            '<input id="__poll_' + uniqueID + '_qid' + numb + '" name="qid' + numb + '" type="hidden" value=""/>' +
            '</div>';

        jq('#__poll_' + uniqueID + '_qbox').append(sb);
        jq("input[name='q" + numb + "']").focus();
        if (numb > 2)
            jq("span[class^='poll_remove_span']").show();
        if (numb >= 15)
            jq('#__poll_' + uniqueID + '_addButton').hide();
    };

    this.RemoveAnswerVariant = function (item) {
        var itemName = jq(item).attr("class");
        var number = parseInt(itemName.substring(17));
        jq(item).parent("div").remove();
        var massVariants = jq("label[class='poll_variantLabel']");
        if (massVariants.length < 3) {
            jq("span[class^='poll_remove_span']").hide();
        }
        var text = massVariants[0].innerHTML;
        var variant = text.substring(0, text.length - 2);
        for (var i = number; i <= massVariants.length; i++) {
            massVariants[i - 1].innerHTML = variant + i + ":";
            jq(massVariants[i - 1]).parent("div").children("span[class^='poll_remove_span']").attr("class", "poll_remove_span_" + i);
            if (i < 10) {
                var textboxs = jq(massVariants[i - 1]).parent("div").children("input");
                jq(textboxs[0]).css("margin-left", "10px");

            }
        }
        var lastIndex = massVariants.length;

        if (lastIndex < 15)
            jq("div[id$='_addButton']").show();
    };
};